[nativert] Move GraphSignature to pytorch core

yiming0416 opened the pull request:
Summary:  
Torch Native Runtime RFC: https://github.com/pytorch/rfcs/pull/72

Added an in-memory representation for input and output specs of a graph. The GraphSignature class models the input and output specs of an exported graph produced by torch.export, which holds the graph information deserialized from the pt2 archive package. Runtime relies on the GraphSignature for weight name lookup and weight loading.

The serialization schema is defined in torch/_export/serde/schema.py 
See more at: https://docs.pytorch.org/docs/stable/export.html#torch.export.ExportGraphSignature

Test Plan: Added tests under `test/cpp/nativert/test_graph_signature.cpp`

Differential Revision: D73895378

Skylion007 commented on torch/nativert/graph/GraphSignature.cpp:
```diff
@@ -0,0 +1,450 @@
+  };
+
+  auto get_valid = [&](const std::vector<std::optional<std::string>>& outputs) {
+    std::vector<std::string> validOutputs;
```
Reserve size here too

Skylion007 commented on torch/nativert/graph/GraphSignature.cpp:
```diff
@@ -0,0 +1,450 @@
+    return validOutputs;
+  };
+
+  VLOG(1) << fmt::format("[{}]", fmt::join(userInputs_, ", "));
```
I think you can use fmt::print here to avoid << operator and temporary strings

yiming0416 replied:
This will only be printed during DEBUG mode. I think `fmt::print` cannot control the verbose level so we need to use `VLOG(1)`?

malfet replied:
You can guard it behind `if (LOG_LEVEL > XYZ)` which would so much faster, as in the current design `fmt::join` will be executed regardless of whether or not logs are actually printed anywhere

yiming0416 replied:
@malfet Thanks for the suggestion! Is `TORCH_CPP_LOG_LEVEL` the correct variable to use here?

yiming0416 replied:
@Skylion007 @malfet  now it uses `fmt::print()` and is guarded under `if (FLAGS_v>0)`

cyyever commented:
We should move `torch/nativert` to `torch/csrc/nativert`

yiming0416 commented:
> We should move `torch/nativert` to `torch/csrc/nativert`. The only C++ files not under `torch/csrc` are
> 
> ```
> ./utils/benchmark/utils/valgrind_wrapper/timer_callgrind_template.cpp
> ./utils/benchmark/utils/valgrind_wrapper/compat_bindings.cpp
> ./utils/benchmark/utils/timeit_template.cpp
> ./_inductor/codegen/aoti_runtime/interface.cpp
> ./lib/libshm_windows/core.cpp
> ./lib/libshm/manager.cpp
> ./lib/libshm/core.cpp
> ./script.h
> ./custom_class.h
> ./library.h
> ./_export/serde/gen-cpp2/export_schema_types_custom_protocol.h
> ./_export/serde/gen-cpp2/export_schema_types.h
> ./_export/serde/gen-cpp2/export_schema_constants.h
> ./extension.h
> ./custom_class_detail.h
> ./_inductor/codegen/cpp_prefix.h
> ./lib/libshm_windows/libshm.h
> ./lib/libshm/libshm.h
> ./lib/libshm/err.h
> ./lib/libshm/alloc_info.h
> ./lib/libshm/socket.h
> ```

based on our earlier discussion, we have decided to put the code under `torch/native`. See conversation in https://github.com/pytorch/pytorch/pull/152953

zhxchen17 commented:
> We should move `torch/nativert` to `torch/csrc/nativert`. The only C++ files not under `torch/csrc` are
> 
> ```
> ./utils/benchmark/utils/valgrind_wrapper/timer_callgrind_template.cpp
> ./utils/benchmark/utils/valgrind_wrapper/compat_bindings.cpp
> ./utils/benchmark/utils/timeit_template.cpp
> ./_inductor/codegen/aoti_runtime/interface.cpp
> ./lib/libshm_windows/core.cpp
> ./lib/libshm/manager.cpp
> ./lib/libshm/core.cpp
> ./script.h
> ./custom_class.h
> ./library.h
> ./_export/serde/gen-cpp2/export_schema_types_custom_protocol.h
> ./_export/serde/gen-cpp2/export_schema_types.h
> ./_export/serde/gen-cpp2/export_schema_constants.h
> ./extension.h
> ./custom_class_detail.h
> ./_inductor/codegen/cpp_prefix.h
> ./lib/libshm_windows/libshm.h
> ./lib/libshm/libshm.h
> ./lib/libshm/err.h
> ./lib/libshm/alloc_info.h
> ./lib/libshm/socket.h
> ```

@cyyever As I mentioned earlier in another PR (https://github.com/pytorch/pytorch/pull/152953#discussion_r2078068928), @malfet requested us to put the code under torch/nativert instead of torch/csrc. 

The context is that torch/csrc historically is intended for Python related code, where nativert is a pure C++ runtime based on torch.export, so they are intended to be on different directories. (Updated RFC: https://github.com/pytorch/rfcs/blob/fc0c9c2fee82828e0bc8eb33a58c05acfabce3a9/RFC-0043-torch-native-runtime.md#proposed-implementation)

swolchok commented on torch/nativert/graph/GraphSignature.h:
```diff
@@ -0,0 +1,139 @@
+
+namespace torch::nativert {
+
+// In memory representation of graph signature.
```
but what is a graph signature?

yiming0416 replied:
@swolchok Added description to this PR and comments in the code

swolchok commented on torch/nativert/graph/GraphSignature.h:
```diff
@@ -0,0 +1,139 @@
+  std::set<std::string> inputNames() const;
+  std::set<std::optional<std::string>> outputNames() const;
+
+  std::unordered_map<std::string, std::string> gradientsToParameters_;
+  std::unordered_map<std::string, std::string> gradientsToUserInputs_;
+  std::unordered_map<std::string, std::string> inputsToParameters_;
+  std::unordered_map<std::string, std::string> inputsToBuffers_;
+  std::unordered_map<std::string, std::string> buffersToMutate_;
+  std::unordered_map<std::string, std::string> inputsToTensorConstants_;
+  std::unordered_map<std::string, std::string> inputsToCustomObjs_;
+  std::unordered_map<std::string, std::string> userInputsToMutate_;
```
this many maps of strings to other strings is an efficiency code smell.

- why do we have multiple parallel maps keyed by the same string? couldn't we, for example, map each input to a struct containing its parameters, buffers, tensor constants, and custom objects?
- is a string really the right data type for all this stuff? 

More context about what is really going on here would be helpful for making concrete suggestions

yiming0416 replied:
@swolchok This representation of `GraphSignature` in `nativert` follows the existing`ExportGraphSignature` (https://docs.pytorch.org/docs/stable/export.html#torch.export.ExportGraphSignature)

yiming0416 replied:
@swolchok Oh I believe there is some misunderstanding. these maps are not keyed by the same string.
As the the `input` in `inputsToParameters_` and `inputsToBuffers_` are not the same inputs.

swolchok commented on torch/nativert/graph/GraphSignature.h:
```diff
@@ -0,0 +1,139 @@
+  // inputsToTensorConstants_
+  std::vector<std::pair<std::string, std::string>> inputsToWeights_;
+
+  std::vector<std::string> parameters_;
+  std::vector<std::string> buffers_;
+  std::vector<std::string> tensorConstants_;
+  std::vector<std::string> customObjs_;
+  std::vector<std::string> nonPersistentBuffers_;
+
+  std::vector<std::string> userInputs_;
+  std::vector<std::string> constantInputs_;
```
here we are duplicating the map keys again! can't we just iterate over the map?

yiming0416 replied:
@swolchok again, these follow the existing `ExportGraphSignature` in `torch.export` (please see my comments above). And later on we rely on these strings for weight look-up and loading. If we don't store them and just iterate over the map, it might add some overhead every time we access them?

swolchok replied:
>  If we don't store them and just iterate over the map, it might add some overhead every time we access them?

if we use a good hash table (such c10::FastMap), I would not expect iterating over map keys to be materially more expensive than iterating over an array.

> these follow the existing ExportGraphSignature

Blindly porting Python data structures to C++ is not appropriate. Python strings are stored by pointer and are often interned (if my sources are to be trusted), whereas C++ std::string is a value type with no interning. Efficiency demands on C++ code are also typically higher; Python is by its nature inefficient and so micro-optimizing Python code and data structures tends to miss the point.

yiming0416 replied:
@swolchok I changed `unordered_set` to `c10::FastSet` and `unordered_map` to `c10::FastMap`
But I have to keep these `std::vector<std::string>`, because we need to maintain the order of these strings as it is an important assumption in nativert for weight loading and some unused weight optimization later.

yiming0416 replied:
> How big is this data structure typically? Could we do something like store std::string_view or std::string* in the maps (except the ones that get touched in replaceAllUses!) since the data is owned by the vectors? (note that this would require appropriate care to make sure the referenced strings don't get moved, such as by vector reallocation)

@swolchok It depends on the model size. For example the size of `parameters_` would be the number of parameters of the model.

yiming0416 replied:
@swolchok I think we can't use `string_view` here because the strings stored in `parameters_` are loaded from an in-memory representation of the serialized json format `torch::_export::GraphSignature` and we free it after `torch::native::GraphSignature` is constructed. So we have to store actual strings.
There was an attempt to dedupe attributes in GraphSignature in D69926343, but it was reverted as it required C++20. I think we can revisit this later when PyTorch build supports C++20

swolchok replied:
> I think we can't use string_view here because the strings stored in parameters_ are loaded from an in-memory representation of the serialized json format torch::_export::GraphSignature and we free it after torch::native::GraphSignature is constructed. So we have to store actual strings.

I meant that the string_view or string* would refer to the string data owned by the std::vectors *in this class*, not in the input.

yiming0416 replied:
> I meant that the string_view or string* would refer to the string data owned by the std::vectors _in this class_, not in the input.

@swolchok I see your points, and I have given that a try. Unfortunately under the current design it's difficult to avoid vector reallocation, so using string_view in the map to refer to strings in the vector will cause downstream weight loading failures. Therefore, we'd prefer to keep the current implementation. We'll definitely resume the attempt to optimize this, as we did in D69926343, when the PyTorch build supports C++20.
Do you think that's okay? Are there any other blocking comments on this PR?

swolchok replied:
>  it's difficult to avoid vector reallocation

AIUI `replaceAllUses` is the only non-const method in this class. I understand why it's difficult to avoid reallocation for the vectors it touches, but it looks like there is a significant subset of the state that is never modified after construction. Why is it difficult to avoid reallocation for that subset?

yiming0416 replied:
> AIUI `replaceAllUses` is the only non-const method in this class. I understand why it's difficult to avoid reallocation for the vectors it touches, but it looks like there is a significant subset of the state that is never modified after construction. Why is it difficult to avoid reallocation for that subset?
@swolchok  Because currently we are dynamically constructing the vector and the map at the same time. And we don't know the size of the vector in advance. So using the string_view of vector in the map may lead to error.
Or are you suggesting we should construct the vector first, and then have another pass to construct the map using the string_view of the vector?

yiming0416 replied:
@swolchok In the latest version I have done the duplication. Now we only have two maps `inputsToWeights_` and `inputsToCustomObjs_` actually saving the strings (these two maps are disjoint). Rest of the maps and vectors will use `string_view` to refer the strings saved in these two maps.

swolchok commented on torch/nativert/graph/GraphSignature.h:
```diff
@@ -0,0 +1,139 @@
+  }
+
+  void lint(
+      const std::set<std::string>& graphInputs,
+      const std::set<std::string>& graphOutputs) const;
```
std::set is not known for its efficiency; using it is concerning

yiming0416 replied:
@swolchok  changed to using `std::unordered_set`

swolchok commented on torch/nativert/graph/GraphSignature.cpp:
```diff
@@ -0,0 +1,449 @@
+}
+
+std::ostream& operator<<(std::ostream& out, const GraphSignature& sig) {
+  out << "inputsToParameters: {\n";
+  for (const auto& [inputName, paramName] : sig.inputsToParameters()) {
+    out << "\t" << inputName << " : " << paramName << std::endl;
+  }
+  out << "}\n";
+
+  out << "inputsToBuffers: {\n";
+  for (const auto& [inputName, bufferName] : sig.inputsToBuffers()) {
+    out << "\t" << inputName << " : " << bufferName << std::endl;
+  }
+  out << "}\n";
+
+  out << "inputsToTensorConstants: {\n";
+  for (const auto& [inputName, tensorConstantName] :
+       sig.inputsToTensorConstants()) {
+    out << "\t" << inputName << " : " << tensorConstantName << std::endl;
+  }
+  out << "}\n";
+
+  return out;
```
why do we print only these three things out of the dozen or so that exist?

yiming0416 replied:
@swolchok Added print for other fields for completeness.

swolchok commented on torch/nativert/graph/GraphSignature.cpp:
```diff
@@ -0,0 +1,449 @@
+
+std::set<std::string> GraphSignature::inputNames() const {
+  std::set<std::string> ret;
+  for (const auto& name : userInputs()) {
+    ret.insert(name);
+  }
+  for (const auto& [inputName, _] : inputsToParameters()) {
+    ret.insert(inputName);
+  }
+  for (const auto& [inputName, _] : inputsToBuffers()) {
+    ret.insert(inputName);
+  }
+  for (const auto& [inputName, _] : inputsToTensorConstants()) {
+    ret.insert(inputName);
+  }
+  for (const auto& [inputName, _] : inputsToCustomObjs()) {
+    ret.insert(inputName);
+  }
```
if we kept our data nicely deduplicated in the first place we wouldn't have to do this expensive operation here

Skylion007 replied:
True, can we not just use an ordered set?

yiming0416 replied:
@Skylion007  do you mind elaborating?

malfet reviewed (changes requested):
Neither RFC nor this PR sheds any light on what is graph signature and why is it needed for native runtime

Skylion007 commented:
> > We should move `torch/nativert` to `torch/csrc/nativert`. The only C++ files not under `torch/csrc` are
> > ```
> > ./utils/benchmark/utils/valgrind_wrapper/timer_callgrind_template.cpp
> > ./utils/benchmark/utils/valgrind_wrapper/compat_bindings.cpp
> > ./utils/benchmark/utils/timeit_template.cpp
> > ./_inductor/codegen/aoti_runtime/interface.cpp
> > ./lib/libshm_windows/core.cpp
> > ./lib/libshm/manager.cpp
> > ./lib/libshm/core.cpp
> > ./script.h
> > ./custom_class.h
> > ./library.h
> > ./_export/serde/gen-cpp2/export_schema_types_custom_protocol.h
> > ./_export/serde/gen-cpp2/export_schema_types.h
> > ./_export/serde/gen-cpp2/export_schema_constants.h
> > ./extension.h
> > ./custom_class_detail.h
> > ./_inductor/codegen/cpp_prefix.h
> > ./lib/libshm_windows/libshm.h
> > ./lib/libshm/libshm.h
> > ./lib/libshm/err.h
> > ./lib/libshm/alloc_info.h
> > ./lib/libshm/socket.h
> > ```
> 
> @cyyever As I mentioned earlier in another PR ([#152953 (comment)](https://github.com/pytorch/pytorch/pull/152953#discussion_r2078068928)), @malfet requested us to put the code under torch/nativert instead of torch/csrc.
> 
> The context is that torch/csrc historically is intended for Python related code, where nativert is a pure C++ runtime based on torch.export, so they are intended to be on different directories. (Updated RFC: https://github.com/pytorch/rfcs/blob/fc0c9c2fee82828e0bc8eb33a58c05acfabce3a9/RFC-0043-torch-native-runtime.md#proposed-implementation)

@malfet, For a variety of reasons (including linting, static analys, etc) it is good practice to have all the C++ code seperated out from the Python code in the repo. There is nothing in the foldername to indicate that any C++ code lives in nativert. I don't think this code will be properly covered by the clang-format, clang-tidy, and other static analysis linters which is a bit dangerous. Also we will need to add these files to the include regex in `.clang-tidy` I would advocate maybe putting these all under some other executor maybe? Doesn't feel right to put them under lib since that would create cyclic dependency.

zhxchen17 commented:
> I don't think this code will be properly covered by the clang-format, clang-tidy, and other static analysis linters which is a bit dangerous.

@Skylion007 In this case should we instead provide enough support for lintrunner and cland-tidy in this case. Happy to add nativert to the tooling coverages.

yiming0416 commented:
> Neither RFC nor this PR sheds any light on what is graph signature and why is it needed for native runtime

@malfet  Added description in the PR as well as comments in the code

yiming0416 commented:
@malfet @swolchok Can you please review again? Thanks!

zhxchen17 commented on test/cpp/nativert/CMakeLists.txt:
```diff
@@ -4,7 +4,9 @@ set(NATIVERT_TEST_ROOT ${TORCH_ROOT}/test/cpp/nativert)
 set(NATIVERT_TEST_SRCS
   ${NATIVERT_TEST_ROOT}/test_tensor_meta.cpp
   ${NATIVERT_TEST_ROOT}/test_mpmc_queue.cpp
+  ${NATIVERT_TEST_ROOT}/test_graph_signature.cpp
   ${TORCH_ROOT}/torch/nativert/graph/TensorMeta.cpp
+  ${TORCH_ROOT}/torch/nativert/graph/GraphSignature.cpp
 )
```
@yiming0416 we updated the test config here such that you don't need to manually add test*.cpp here. maybe you want to rebase?

yiming0416 replied:
@zhxchen17 just rebased.

swolchok commented on torch/nativert/graph/GraphSignature.cpp:
```diff
@@ -0,0 +1,505 @@
+void checkInputOrders(
+    const std::vector<torch::_export::InputSpec>& inputSpecs) {
+  // Map each tag to its index in the expected order
+  std::unordered_map<torch::_export::InputSpec::Tag, size_t> tagOrderMap = {
+      {torch::_export::InputSpec::Tag::TOKEN, 0},
+      {torch::_export::InputSpec::Tag::PARAMETER, 1},
+      {torch::_export::InputSpec::Tag::BUFFER, 2},
+      {torch::_export::InputSpec::Tag::TENSOR_CONSTANT, 3},
+      {torch::_export::InputSpec::Tag::CUSTOM_OBJ, 4}};
```
1) creating a new hash table on each call to this function is inefficient
2) this is very small; linear search on a static constexpr array of std::pair would very likely be fine
3) you don't need to use a whole size_t for the value. uint32_t would be fine

swolchok commented on torch/nativert/graph/GraphSignature.cpp:
```diff
@@ -0,0 +1,505 @@
+    return;
+  }
+
+  std::string errorMsg =
+      "Error: Value name difference detected between graph signature and graph nodes:\n";
+  errorMsg += "Signature value names:\n";
+  errorMsg += fmt::format("[{}]\n", fmt::join(sigNames, ", "));
+  errorMsg += "Graph node names:\n";
+  errorMsg += fmt::format("[{}]", fmt::join(graphNames, ", "));
```
nit: why not use fmt::format once to construct the whole message? iteratively concatenating is inefficient

swolchok commented on torch/nativert/graph/GraphSignature.cpp:
```diff
@@ -0,0 +1,505 @@
+  if (it == map.end()) {
+    return;
+  }
+  std::string value = it->second;
```
std::move

swolchok commented on torch/nativert/graph/GraphSignature.cpp:
```diff
@@ -0,0 +1,505 @@
+  }
+  std::string value = it->second;
+  map.erase(it);
+  map.emplace(replacement, value);
```
std::move

swolchok commented on torch/nativert/graph/GraphSignature.cpp:
```diff
@@ -0,0 +1,509 @@
+        inputSpec.tag() == torch::_export::InputSpec::Tag::CONSTANT_INPUT) {
+      continue;
+    }
+    uint32_t tagIndex = static_cast<uint32_t>(-1);
+    for (const auto& [tag, index] : tagOrderArray) {
+      if (tag == inputSpec.tag()) {
+        tagIndex = index;
+        break;
+      }
+    }
+    TORCH_CHECK(
+        tagIndex != static_cast<uint32_t>(-1),
+        "Unknown InputSpec tag encountered.");
```
it would probably be simpler to use std::find for this after switching to std::array

swolchok commented on torch/nativert/graph/GraphSignature.cpp:
```diff
@@ -0,0 +1,509 @@
+void checkInputOrders(
+    const std::vector<torch::_export::InputSpec>& inputSpecs) {
+  // Map each tag to its index in the expected order
+  static constexpr std::pair<torch::_export::InputSpec::Tag, uint32_t>
+      tagOrderArray[] = {
+          {torch::_export::InputSpec::Tag::TOKEN, 0},
+          {torch::_export::InputSpec::Tag::PARAMETER, 1},
+          {torch::_export::InputSpec::Tag::BUFFER, 2},
+          {torch::_export::InputSpec::Tag::TENSOR_CONSTANT, 3},
+          {torch::_export::InputSpec::Tag::CUSTOM_OBJ, 4}};
```
use std::array. there should be a linter that will complain about this too

Skylion007 commented on torch/nativert/graph/GraphSignature.cpp:
```diff
@@ -0,0 +1,511 @@
+    default:
+      TORCH_CHECK(false, "Unknown InputSpec tag encountered.");
+  }
+  return std::make_pair(argName, tagName);
```
Nit: std::move both these strings

Skylion007 commented on torch/nativert/graph/GraphSignature.cpp:
```diff
@@ -0,0 +1,511 @@
+      }
+      case torch::_export::InputSpec::Tag::CUSTOM_OBJ: {
+        customObjs_.push_back(inputSpec.get_custom_obj().get_custom_obj_name());
+        inputsToCustomObjs_.insert(
```
Does the backing class here not have try_emplace?

Skylion007 commented on torch/nativert/graph/GraphSignature.cpp:
```diff
@@ -0,0 +1,511 @@
+}
+
+c10::FastSet<std::string> GraphSignature::inputNames() const {
+  c10::FastSet<std::string> ret;
```
Both backing implementations of FastSet supports reserve.

yiming0416 commented:
@Skylion007 @swolchok  I have addressed most of the comments. Except:
1. We'll have to keep std::vector<std::string> for parameters_, buffers_, etc as we need to keep the order of their names in the graph signature, and this is an assumption that downstream weight processing relies on.
2. We'd prefer to keep separate mappings like inputsToParameters_ and inputsToBuffers_ , as they will be consumed separately in the runtime.
But if these are indeed blocking comments, I'll be happy to further discuss

yiming0416 commented:
hi @swolchok can you please take a look at my response above and review again? Thanks!

yiming0416 commented:
@swolchok Sorry for pinging again but can you please take a look at my response above and let me know if there is any other blocking comments? Thanks!

swolchok commented on torch/nativert/graph/GraphSignature.cpp:
```diff
@@ -0,0 +1,479 @@
+std::pair<std::string, std::string> getSpecDetails(
+    const torch::_export::InputSpec& inputSpec) {
+  // Retrieve the argument name and spec tag name
+  std::string argName;
+  std::string tagName;
```
nit: this would probably be more readable if you just returned a std::make_pair from each switch case rather than assigning to argName and tagName and then falling down to a single point of return.

swolchok commented on torch/nativert/graph/GraphSignature.cpp:
```diff
@@ -0,0 +1,479 @@
+    currentOrderIndex = tagIndex;
+    // Additional check for buffers
+    if (inputSpec.tag() == torch::_export::InputSpec::Tag::BUFFER) {
+      if (!inputSpec.get_buffer().get_persistent()) {
+        seenNonPersistentBuffer = true;
+      } else if (
+          inputSpec.get_buffer().get_persistent() && seenNonPersistentBuffer) {
```
why do you need to repeat the check of `inputSpec.get_buffer().get_persistent()`? couldn't you just write ```
else {
    TORCH_CHECK(!seenNonPersistentBuffer, "Persistent buffers must come before non-persistent buffers.");
}
```

yiming0416 committed:
[nativert] Move GraphSignature to pytorch core (#152969)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/152969

An in-memory representation of `GraphSignature` which will be consumed by the runtime.

Test Plan: Added tests under `test/cpp/nativert/test_graph_signature.cpp`

Reviewed By: zhxchen17

Differential Revision: D73895378

yiming0416 commented:
@pytorchbot merge -f "landed internally"
